apps/src/codebridge/FileBrowser/index.tsx

```diff
@@ -308,55 +320,42 @@ export const FileBrowser = React.memo(() => {
             message = codebridgeI18n.duplicateSupportFileError({fileName});
           }
         }
-        if (message) {
-          dialogControl?.showDialog({
-            type: DialogType.GenericAlert,
-            title: message,
-          });
-          return true;
-        } else {
-          return false;
-        }
-      },
-    [dialogControl]
-  );
-
-  const innerNewFolder: newFolderType = useMemo(
-    () => (parentId, folderName) => {
-      const folderId = getNextFolderId(Object.values(project.folders));
-
-      if (!folderName) {
-        return;
-      }
-
-      const existingFolder = Object.values(project.folders).some(
-        f => f.name === folderName && f.parentId === parentId
-      );
-      if (existingFolder) {
-        dialogControl?.showDialog({
-          type: DialogType.GenericAlert,
-          title: codebridgeI18n.folderExistsError(),
-        });
-        return;
-      }
 
-      newFolder({parentId, folderName, folderId});
-    },
-    [newFolder, project.folders, dialogControl]
+        return message;
+      },
+    []
   );
 
   const newFolderPrompt: FilesComponentProps['newFolderPrompt'] = useMemo(
     () =>
-      (parentId = DEFAULT_FOLDER_ID) => {
-        dialogControl?.showDialog({
+      async (parentId = DEFAULT_FOLDER_ID) => {
+        const results = await dialogControl.showDialog({
           type: DialogType.GenericPrompt,
           title: codebridgeI18n.newFolderPrompt(),
-          handleConfirm: folderName => {
-            innerNewFolder(parentId, folderName);
+          validateInput: (folderName: string) => {
+            if (!folderName.length) {
+              return;
+            }
+            if (!validateName(folderName)) {
+              return codebridgeI18n.invalidNameError();
+            }
+            const existingFolder = Object.values(project.folders).some(
+              f => f.name === folderName && f.parentId === parentId
+            );
+            if (existingFolder) {
+              return codebridgeI18n.folderExistsError();
+            }
           },
         });
+        if (results.type !== 'confirm') {
+          return;
+        }
+        const folderName = extractPrompt(results);
+
+        const folderId = getNextFolderId(Object.values(project.folders));
+        newFolder({parentId, folderName, folderId});
       },
-    [dialogControl, innerNewFolder]
+    [dialogControl, newFolder, project.folders]
```

Comment:
This style of mod is repeated through the rest of the file. The gist is:'

- Get rid of the additional `inner*` function that was called in the `handleConfirm` call back
- Move any error handling from the `inner*` function to the `validateInput` function to display errors inline.
- Now async. Await the dialogControl's promise. Assuming confirm was clicked, extract the prompted value and then take the appropriate action.

Reply:
Most of these functions also have more explicit and better error handling, such as enforcing naming or clearly displaying errors that may have occurred. There were a few cases where things could transparently sneak by before.